pkg/storegateway/indexheader/index/postings.go

```diff
@@ -0,0 +1,514 @@
+			// Let's exit and do binary search again / exit if nothing wanted.
+
+			if len(newSameRngs) > 0 {
+				// We added some ranges in this iteration. Use next posting offset as the end of our ranges.
+				// We know it exists as we never go further in this loop than e.offsets[i, i+1].
+
+				skipNAndName(&d, &buf)
+				d.UvarintBytes() // Label value.
+				postingOffset := int64(d.Uvarint64())
+
+				for j := range newSameRngs {
+					newSameRngs[j].End = postingOffset - crc32.Size
+				}
+				rngs = append(rngs, newSameRngs...)
+			}
+			break
+		}
+		if d.Err() != nil {
+			return nil, errors.Wrap(d.Err(), "get postings offset entry")
+		}
+	}
+
+	return rngs, nil
+}
+
+func (t *PostingOffsetTableV2) LabelValues(name string, filter func(string) bool) ([]string, error) {
+	e, ok := t.postings[name]
+	if !ok {
+		return nil, nil
+	}
+	if len(e.offsets) == 0 {
+		return nil, nil
+	}
+	values := make([]string, 0, len(e.offsets)*t.postingOffsetsInMemSampling)
+
+	// Don't Crc32 the entire postings offset table, this is very slow
+	// so hope any issues were caught at startup.
+	d := t.factory.NewDecbufAtUnchecked(t.tableOffset)
+	defer t.factory.Close(d)
+
+	d.Skip(e.offsets[0].tableOff)
+	lastVal := e.offsets[len(e.offsets)-1].value
+
+	skip := 0
+	for d.Err() == nil {
+		if skip == 0 {
+			// These are always the same number of bytes,
+			// and it's faster to skip than parse.
+			skip = d.Len()
+			d.Uvarint()      // Keycount.
+			d.UvarintBytes() // Label name.
+			skip -= d.Len()
+		} else {
+			d.Skip(skip)
+		}
+		s := yoloString(d.UvarintBytes()) // Label value.
+		if filter == nil || filter(s) {
+			// Clone the yolo string since its bytes will be invalidated as soon as
+			// any other reads against the decoding buffer are performed.
+			values = append(values, strings.Clone(s))
+		}
+		if s == lastVal {
+			break
+		}
+		d.Uvarint64() // Offset.
+	}
+	if d.Err() != nil {
+		return nil, errors.Wrap(d.Err(), "get postings offset entry")
+	}
+	return values, nil
+}
+
+func (t *PostingOffsetTableV2) LabelNames() ([]string, error) {
+	labelNames := make([]string, 0, t.LabelNameCount())
+
+	err := t.ForEachLabelName(func(name string) error {
+		labelNames = append(labelNames, name)
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	sort.Strings(labelNames)
+
+	return labelNames, nil
+}
+
+func (t *PostingOffsetTableV2) ForEachLabelName(f func(name string) error) error {
+	allPostingsKeyName, _ := index.AllPostingsKey()
+
+	for name := range t.postings {
+		if name == allPostingsKeyName {
+			continue
+		}
+
+		if err := f(name); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (t *PostingOffsetTableV2) LabelNameCount() int {
+	// This count might include AllPostingsKey's name, but that's OK - we use this method
+	// to pre-allocate slices to hold all label names, and slightly over-allocating is acceptable
+	// in this case.
+	return len(t.postings)
+}
```

Comment:
In which cases it would not include the `AllPostingsKey`? I think it's always present in the IndexV2 format, right?

If I'm wrong, I'd rename the method to `EstimateLabelNameCount`.

Reply:
Yep, you're correct, it should always be included in v2: https://github.com/grafana/mimir/blob/c44a07dd4c4e38a19ba01d676555c3c84d30b461/vendor/github.com/prometheus/prometheus/tsdb/index/index.go#L853

I've updated the comment to reflect this.

Is `AllPostingsKey` also included in v1?